### PR TITLE
feat(mcp): analyze changes after `do`

### PR DESCRIPTION
**Pull request by p0deje**
This PR improves MCP `do` tool to automatically calculate the difference between before and after accessibility tree, analyze the diff and give the summary of the differences back to the MCP client. It lets users receive feedback about what changed in the UI after executing an action. Prelimiary results show that it increases MCP clients accuracy because they can understand the effect of their commands and decide what to do next immediately after issuing `do` tool call. This effectively allows general-purpose agents to act like a browser agent with one-shot prompts:

<details>
<summary>Example of using Claude Code</summary>

```
❯ You are a browser agent for https://arxiv.org. Search for the latest preprints about 'quantum computing'.

⏺ I'll help you search for quantum computing preprints on arXiv. Let me break this down into steps and get started.                                                                                                                            11:43 AM  claude-sonnet-4-5-20250929

⏺ alumnium - start_driver (MCP)(capabilities: "{\"platformName\": \"chrome\"}")
  ⎿  Chrome driver started successfully (driver_id: python-1769024610)
     Model: azure_openai/gpt-5-nano

⏺ alumnium - do (MCP)(driver_id: "python-1769024610", goal: "navigate to https://arxiv.org")
  ⎿ {
      "explanation": "Goal: navigate to https://arxiv.org. Since the only available focusable element is the root web area with a current URL, the direct action is to navigate that element to the specified URL.",
      "performed_steps": [
        {
          "name": "navigate to \"https://arxiv.org\" URL",
          "tools": [
    "NavigateToUrlTool(url=https://arxiv.org)"
          ]
        }
      ],
      "ui_changes": "URL changed to https://arxiv.org/. A lot of initial navigation content was added by loading the full arXiv.org e-Print archive homepage. The page now includes a prominent header region, a skip-to-content link, branding, login, and a full search bar with a
    field selector. The main content was expanded into extensive subject sections (Physics, Mathematics, Computer Science, Quantitative Biology, Quantitative Finance, Statistics, Electrical Engineering and Systems Science, Economics) with nested topic lists and quick links
    (new/recent/search) for each area. There is also an About section, a detailed description paragraph about arXiv, and a secondary footer/navigation region with About, Help, Contact, Subscribe, Privacy, and Accessibility items."
    }

⏺ alumnium - do (MCP)(driver_id: "python-1769024610", goal: "search for \"quantum computing\"")
  ⎿ {
      "explanation": "Goal: search for \"quantum computing\" on the arXiv page. I will locate the text input labeled \"Search term or terms\" and enter the query, then trigger the search by clicking the nearby \"Search\" button in the same search area. This directly corresponds
     to the input field and action controls present in the accessibility tree under the main search region.",
      "performed_steps": [
        {
          "name": "type \"quantum computing\" to textbox \"Search term or terms\"",
          "tools": [
    "TypeTool(id=32, text=quantum computing)"
          ]
        },
        {
          "name": "click button \"Search\"",
          "tools": [
    "ClickTool(id=62)"
          ]
        }
      ],
      "ui_changes": "URL changed to https://arxiv.org/search/?query=quantum+computing&searchtype=all&source=header. A page navigation change occurred: the UI moved from a broad arXiv homepage to a full search-results experience for “quantum computing.” The after diff replaces
    the simple root with a header/banner, a skip-to-content link, and a prominent arXiv logo, then expands into a complete search UI and results area rather than a static intro. Key content changes include a live search bar and advanced-search links, per-page and sort controls,
     a long list of arXiv results with each item showing titles, authors, abstracts, and links (abs/pdf/other), and a pagination/navigation block for moving through results. The structure now includes a main content region with a “Showing 1–50 of … results” heading, a dedicated
     results list, and a secondary navigation section for status, help, and policy links."
    }


⏺ alumnium - get (MCP)(driver_id: "python-1769024610", data: "list of the first 10 search results including title, authors, submission date, and abstract")
  ⎿ ['1) Device-independent quantum memory certification in two-point measurement experiments', 'Authors: Leonardo S. V. Santos; Peter Tirler; Michael Meth; Lukas Gerster; Manuel John; Keshav Pareek; Tim Gollerthan; Martin Ringbauer; Otfried Gühne', 'Submitted 20 January,
    2026', 'Abstract: Quantum memories are key components of emerging…\n\n2) Quantum mixing on large Schreier graphs', 'Authors: Charles Bordenave; Cyril Letrouit; Mostafa Sabri', 'Submitted 20 January, 2026', 'Abstract: Quantum ergodicity describes the delocalization of most
    eigenfunctions of Laplace-type operators on graphs or manifolds exhibiting chaotic classical dynamics.…\n\n3) Sharp Inequalities for Schur-Convex Functionals of Partial Traces over Unitary Orbits', 'Authors: Pablo Costa Rico; Pavel Shteyner', 'Submitted 20 January, 2026',
    'Abstract: While many bounds have been proved for partial trace inequalities over the last decades for a large variety of quantities, recent problems in quantum information theory demand sharper bounds. In this work, we study optimal bounds for partial trace quantities in
    terms of the spectrum; equivalently, we determine the best bounds attainable over unitary orbits o…\n\n4) Measurement of the Z production cross section and search for anomalous neutral triple gauge couplings in pp collisions at  = 13 TeV', 'Authors: CMS Collaboration',
    'Submitted 20 January, 2026', 'Abstract: Measurements of the production cross section and limits on anomalous couplings…\n\n5) Software Testing in the Quantum World', 'Authors: Rui Abreu; Shaukat Ali; Paolo Arcaini; Jose Campos; Michael Felderer; Claude Gravel; Fuyuki
    Ishikawa; Stefan Klikovits; Andriy Miranskyy; Mohammad Mousavi; Masaomi Yamaguchi; Lei Zhang; Jianjun Zhao; Anila Mjeda', 'Submitted 20 January, 2026', 'Abstract: Quantum software testing is essential for ensuring reliability and correctness of quantum programs; this paper
    discusses testing strategies and challenges in the quantum domain…\n\n6) Optimal Construction of Two-Qubit Gates using the Symmetries of B Gate Equivalence Class', 'Authors: Henryk Gzyl; (additional authors listed in the entry)', 'Submitted 20 January, 2026', 'Abstract:
    …local equivalence classes on these planes, with and without the B gate equivalence class, such that each of them can be used to construct a parameterized universal two-qubit quantum circuit that involves only two nonlocal two-qubit gates. We also discuss the implementation
     of the gates from a few families of local equivalence classes on superconducting…\n\n7) Tensor Network Assisted Distributed Variational Quantum Algorithm for Large Scale Combinatorial Optimization Problem', 'Authors: Yuhan Huang; Siyuan Jin; Yichi Zhang; Qi Zhao; Jun Qi;
    Qiming Shao; (additional authors listed in the entry)', 'Submitted 20 January, 2026', 'Abstract: Although quantum…', '8) Optimal Construction of Two-Qubit Gates using the Symmetries of B Gate Equivalence Class (duplicate of 6 listed here for context)', 'Authors: Henryk
    Gzyl; (additional authors listed in the entry)', 'Submitted 20 January, 2026', 'Abstract: …local equivalence classes on these planes…\n\n9) On spooky action at a distance and conditional probabilities', 'Authors: Henryk Gzyl; (additional authors listed in the entry)',
    'Submitted 20 January, 2026', 'Abstract: The aim of this exposé is to make explicit the analogy between the classical notion of non-independent probability distribution and the quantum notion of entangled state.…\n\n10) Inverse Area Corrections to Black Hole Entropy Area
    Formula in F(R) Gravity and Gravitational Wave Observations', 'Authors: Rohit Das; Parthasarathi Majumdar; Debadrita Mukherjee', 'Submitted 20 January, 2026', 'Abstract: …results validating the Hawking Area Theorem for binary black hole coalescences, implies constraints on
    parameters of F(R) gravity. For the sake of comparison, we present a computation of inverse area corrections for…']


⏺ alumnium - stop_driver (MCP)(driver_id: "python-1769024610")
  ⎿  Driver python-1769024610 closed.
     Artifacts saved to: /Users/p0deje/Development/alumnium/packages/python/tmp/alumnium/python-1769024610
     Token usage statistics:
     - Total: 85718 tokens (80798 input, 4920 output)
     - Cached: 0 tokens (0 input, 0 output)

∴ Thinking…

  Great! I've successfully completed the task and closed the browser driver. Now let me present the search results to the user in a clear and organized format.

⏺ I successfully searched arXiv for quantum computing preprints. Here are the latest results (all submitted January 20, 2026):                                                                                                                 11:45 AM  claude-sonnet-4-5-20250929

  Latest Quantum Computing Preprints on arXiv

  1. Device-independent quantum memory certification in two-point measurement experiments
  - Authors: Leonardo S. V. Santos, Peter Tirler, Michael Meth, Lukas Gerster, Manuel John, Keshav Pareek, Tim Gollerthan, Martin Ringbauer, Otfried Gühne
  - Submitted: January 20, 2026
  - Abstract: Quantum memories are key components of emerging quantum technologies...

  2. Quantum mixing on large Schreier graphs
  - Authors: Charles Bordenave, Cyril Letrouit, Mostafa Sabri
  - Submitted: January 20, 2026
  - Abstract: Quantum ergodicity describes the delocalization of most eigenfunctions of Laplace-type operators on graphs or manifolds exhibiting chaotic classical dynamics...

  3. Sharp Inequalities for Schur-Convex Functionals of Partial Traces over Unitary Orbits
  - Authors: Pablo Costa Rico, Pavel Shteyner
  - Submitted: January 20, 2026
  - Abstract: Studies optimal bounds for partial trace quantities in quantum information theory...

  4. Measurement of the Z production cross section and search for anomalous neutral triple gauge couplings in pp collisions at √s = 13 TeV
  - Authors: CMS Collaboration
  - Submitted: January 20, 2026
  - Abstract: Measurements of production cross sections and limits on anomalous couplings...

  5. Software Testing in the Quantum World
  - Authors: Rui Abreu, Shaukat Ali, Paolo Arcaini, et al.
  - Submitted: January 20, 2026
  - Abstract: Discusses quantum software testing strategies and challenges for ensuring reliability and correctness of quantum programs...

  6. Optimal Construction of Two-Qubit Gates using the Symmetries of B Gate Equivalence Class
  - Authors: Henryk Gzyl, et al.
  - Submitted: January 20, 2026
  - Abstract: Explores local equivalence classes for constructing parameterized universal two-qubit quantum circuits with implementation on superconducting systems...

  7. Tensor Network Assisted Distributed Variational Quantum Algorithm for Large Scale Combinatorial Optimization Problem
  - Authors: Yuhan Huang, Siyuan Jin, Yichi Zhang, Qi Zhao, Jun Qi, Qiming Shao, et al.
  - Submitted: January 20, 2026
  - Abstract: Addresses quantum approaches to large-scale combinatorial optimization...

  8. On spooky action at a distance and conditional probabilities
  - Authors: Henryk Gzyl, et al.
  - Submitted: January 20, 2026
  - Abstract: Explores the analogy between classical non-independent probability distributions and quantum entangled states...

  9. Inverse Area Corrections to Black Hole Entropy Area Formula in F(R) Gravity and Gravitational Wave Observations
  - Authors: Rohit Das, Parthasarathi Majumdar, Debadrita Mukherjee
  - Submitted: January 20, 2026
  - Abstract: Validates the Hawking Area Theorem for binary black hole coalescences with constraints on F(R) gravity parameters...

  All these preprints were submitted today and represent cutting-edge research across quantum computing, quantum information theory, quantum algorithms, and related physics applications.
```

</details>

On top of that, this PR adds a couple of improvements to prompt organization (removing duplicated prompts) and more  precise explanation of `do` tool in MCP (instructing to use it with a single-step goal).

### CODE REVIEW

File: packages/python/src/alumnium/clients/http_client.py [hunk 1]
```diff
@@ -112,6 +112,29 @@ def find_element(self, description: str, accessibility_tree: str) -> dict:
         response.raise_for_status()
         return response.json()["elements"][0]
 
+    def analyze_ui_changes(
```
Comment (sai-meyyappan) [inline packages/python/src/alumnium/clients/http_client.py]: If the UI analysis fails (LLM timeout, rate limit etc ), does the do action still succeed?

File: packages/python/src/alumnium/server/accessibility/accessibility_tree_diff.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    after_xml: str
+    _diff: str = field(init=False, default="")
+
+    def compute(self) -> str:
+        if not self._diff:
+            self._diff = self._format_as_git_diff()
+
+        return self._diff
+
+    def _format_as_git_diff(self) -> str:
```
Comment (sai-meyyappan) [inline packages/python/src/alumnium/server/accessibility/accessibility_tree_diff.py:17]: Just want to confirm that this works for large accessibility trees?

File: packages/python/src/alumnium/server/accessibility/server_chromium_accessibility_tree.py [hunk 3]
```diff
@@ -80,8 +79,13 @@ def _xml_to_node(self, elem: Element) -> dict:
 
         return node
 
-    def to_xml(self):
-        """Converts the nested tree to XML format using role.value as tags."""
+    def to_xml(self, exclude_attrs: set[str] | None = None):
+        """Converts the nested tree to XML format using role.value as tags.
+
+        Args:
+            exclude_attrs: Optional set of attribute names to exclude from output.
```
Comment (sai-meyyappan) [inline packages/python/src/alumnium/server/accessibility/server_chromium_accessibility_tree.py:86]: Thoughts on making `exclude_attrs` configurable at the session/client level?

**Review by sai-meyyappan — APPROVED** [review]
looks good overall. Just a few questions.

File: packages/python/tests/server/accessibility/test_accessibility_tree_diff.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
+from alumnium.server.accessibility import AccessibilityTreeDiff
+
+
```
Comment (sai-meyyappan) [inline packages/python/tests/server/accessibility/test_accessibility_tree_diff.py:21]: Maybe a few more edge case tests here - 

- Empty diff when trees are identical
- Full tree shown as additions
- Full tree shown as deletions
- Verify behavior with very large diffs
- Verify exclude_attrs filters correctly

File: packages/python/src/alumnium/clients/http_client.py (hunk 1 above)
Comment (p0deje) [reply packages/python/src/alumnium/clients/http_client.py]: Good point, `do` should still succeed without providing `ui_changes` field.

File: packages/python/src/alumnium/server/accessibility/accessibility_tree_diff.py (hunk 2 above)
Comment (p0deje) [reply packages/python/src/alumnium/server/accessibility/accessibility_tree_diff.py:17]: It does, I'll add some tests to demonstrate that.

File: packages/python/src/alumnium/server/accessibility/server_chromium_accessibility_tree.py (hunk 3 above)
Comment (p0deje) [reply packages/python/src/alumnium/server/accessibility/server_chromium_accessibility_tree.py:86]: I think it's fine at the moment, we mostly care about randomized attributes like `id`, while other attributes have certain meaning.

File: packages/python/tests/server/accessibility/test_accessibility_tree_diff.py (hunk 4 above)
Comment (p0deje) [reply packages/python/tests/server/accessibility/test_accessibility_tree_diff.py:21]: Good catch, will add those!
